### PR TITLE
Entry point api

### PR DESCRIPTION
**Pull request by zsluedem**
Introduce a wrapped entry point api that handle the error in our own way.

### CODE REVIEW

File: src/contracts/gen.rs [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+);
+
+abigen!(Account, "$OUT_DIR/IAccount.sol/IAccount.json");
+
+abigen!(
+    UserOperation,
+    "$OUT_DIR/UserOperation.sol/UserOperationLib.json"
+);
+
+impl From<UserOp> for entry_point_api::UserOperation {
```
Comment (zsluedem) [inline src/contracts/gen.rs:31]: This is dup to the pr #24  . Just want to put it here to compile well. Should be removed later.

File: src/contracts/entrypoint.rs [hunk 2]
```diff
@@ -0,0 +1,238 @@
...
+}
+
+#[derive(Debug)]
+pub enum SimulateResult {
+    SimulationResult(entry_point_api::SimulationResult),
+    SimulationResultWithAggregation(entry_point_api::SimulationResultWithAggregation),
+}
+
+#[derive(Debug, Deserialize, PartialEq)]
+pub struct JsonRpcError {
```
Comment (zsluedem) [inline src/contracts/entrypoint.rs:175]: This is actually the same as https://github.com/gakonst/ethers-rs/blob/master/ethers-providers/src/transports/common.rs#L14-L22. But sadly, ethers-rs doesn't expose this type.

File: src/contracts/entrypoint.rs [hunk 3]
```diff
@@ -0,0 +1,238 @@
...
+    pub message: String,
+    /// Additional data
+    pub data: Option<String>,
+}
+
+impl FromStr for JsonRpcError {
+    type Err = anyhow::Error;
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let re = Regex::new(
+            r###"code: (\d+), message: ([^,]*), data: (None|Some\(String\("([^)]*)"\))"###,
```
Comment (zsluedem) [inline src/contracts/entrypoint.rs:188]: https://github.com/gakonst/ethers-rs/blob/master/ethers-providers/src/transports/common.rs#L26
Revert back to JsonRpcError based on the `Display`.

File: src/contracts/entrypoint.rs [hunk 4]
```diff
@@ -0,0 +1,238 @@
...
+            }
+        }
+    }
+
+    async fn handle_aggregated_ops<U: Into<UserOperation>>(
+        &self,
+        ops_per_aggregator: Vec<U>,
+        beneficiary: Address,
+    ) -> Result<(), EntryPointErr> {
+        todo!()
```
Comment (zsluedem) [inline src/contracts/entrypoint.rs:156]: I haven't found good test cases for this. I would make it work in another pr.

File: src/contracts/entrypoint.rs [hunk 5]
```diff
@@ -0,0 +1,238 @@
...
+
+#[cfg(test)]
+mod test {
+    use ethers::types::Bytes;
+
+    use super::JsonRpcError;
+    use std::str::FromStr;
+
+    #[test]
+    fn json_rpc_errpr_parse() {
```
Comment (Vid201) [inline src/contracts/entrypoint.rs:213]: `json_rpc_err_parse`?

File: src/contracts/gen.rs [hunk 6]
```diff
@@ -0,0 +1,47 @@
+use crate::types::user_operation::UserOperation as UserOp;
+use ethers::contract::abigen;
+abigen!(
```
Comment (Vid201) [inline src/contracts/gen.rs:3]: Should we comment out contracts we don't need?

File: src/contracts/entrypoint.rs (hunk 2 above)
Comment (Vid201) [reply src/contracts/entrypoint.rs:175]: Do you know if there is any specific reason why that type is not exposed? Maybe we should update ethers-rs to expose it.

File: src/contracts/entrypoint.rs [hunk 7]
```diff
@@ -0,0 +1,238 @@
...
+#[derive(Debug)]
+pub enum EntryPointErr {
+    FailedOp(entry_point_api::FailedOp),
+    NetworkErr, // TODO
+    DecodeErr(String),
+    UnknownErr(String), // describe impossible error. We should fix the codes here(or contract codes) if this occurs.
+}
+
+#[derive(Debug)]
+pub enum SimulateResult {
```
Comment (Vid201) [inline src/contracts/entrypoint.rs:169]: SimulateValidationResult sounds better to me (more precise based on the SC's function name

File: src/contracts/entrypoint.rs [hunk 8]
```diff
@@ -0,0 +1,238 @@
...
+                AbiDecode::decode_hex(data).map_err(|_| {
+                    EntryPointErr::DecodeErr(format!(
+                        "{:?} data field could not be deserialize to EntryPointAPIErrors",
+                        err_msg
+                    ))
+                })
+            })
+    }
+
+    pub async fn get_user_op_hash<U: Into<UserOperation>>(
```
Comment (Vid201) [inline src/contracts/entrypoint.rs:64]: I do not think this function is needed; it's implemented in user_operation.rs (and no need to make another RPC call to provider/client).

**Review by Vid201 — CHANGES_REQUESTED** [review]

**Comment by Vid201** [comment]
This looks good; good job! Left some comments

File: src/contracts/entrypoint.rs [hunk 9]
```diff
@@ -0,0 +1,238 @@
...
+        Ok(JsonRpcError {
+            code,
+            message: message.to_string(),
+            data,
+        })
+    }
+}
+
+#[cfg(test)]
+mod test {
```
Comment (Vid201) [inline src/contracts/entrypoint.rs:206]: `mod tests`

File: src/contracts/entrypoint.rs (hunk 2 above)
Comment (zsluedem) [reply src/contracts/entrypoint.rs:175]: Besides JsonRpcError, there are a lot of middleware error types in ethers-rs. What kind of error you get depends on the middleware and it is dynamic dispatch. The ethers-rs team told me it is going to change in version 2. I didn't explore how difficult to fix it in the current version.

FYI, the current solution is not handling all the situations well. I haven't met any network errors (or middleware error) yet. And we need to improve this in the future.

Comment (Vid201) [reply src/contracts/entrypoint.rs:168]: Ok, got it, thanks for the explanation

**Review by Vid201 — APPROVED** [review]

**Comment by Vid201** [comment]
@zsluedem It's ready for merge unless you plan to make any additional changes.

**Comment by zsluedem** [comment]
@Vid201 Thanks
